app/features/connect_with_them/tools.py

```diff
@@ -0,0 +1,137 @@
+        data = response.json()['data']['children']
+        subreddits = [post['data']['display_name'] for post in data]
+        return f"These are the subreddits related to '{topic}': {', '.join(subreddits)}"
+    except RequestException as e:
+        print(f"An error occurred during the request: {e}")
+        return f"No subreddits found for '{topic}'. Error: {str(e)}"
+
+def scrape_page(category: str, num_articles: int) -> list:
+    # Build the URL based on the category
+    url = f"https://www.edutopia.org/{categories_mapping[category]}"
+
+    try:
+        # Send a GET request to fetch the page content
+        response = requests.get(url)
+        response.raise_for_status()  # Raise an exception for HTTP errors
+
+        # Parse the HTML content using BeautifulSoup
+        soup = BeautifulSoup(response.text, 'html.parser')
+
+        # Find the 'main' section of the page
+        main = soup.find('main')
+        if not main:
+            print("Main section not found")
+            return []
+
+        # Find the 'ul' element within the 'main' section
+        ul = main.find('ul')
+        if not ul:
+            print("List (ul) not found")
+            return []
+
+        # Find all 'li' elements within the 'ul'
+        li_tags = ul.find_all('li')
+
+        # Extract the 'href' attribute from 'a' tags within 'li'
+        article_links = [li.find('a')['href'] for li in li_tags if li.find('a')]
+
+        # Return the required number of article links
+        return [f'https://www.edutopia.org{link}' for link in article_links[:min(num_articles, len(article_links))]]
+
+    except requests.exceptions.RequestException as e:
+        print(f"Request error: {e}")
+        return []
+
+def Search_Articles(category: str) -> str:
+    try:
+        if category not in list(categories_mapping.keys()) :
+            print(f"\n category before changing : {category}")
+            if "'" in category :
+                category = category.replace("'","")
+            if '"' in category :
+                category = category.replace('"','')
+            print(f"\n category after changing : {category}")
+            if category not in list(categories_mapping.keys()) :
+                return f"Invalid category: {category}. Valid categories are: {list(categories_mapping.keys())}"
+
+        links = scrape_page(category, 3)
+        if not links:
+            return f"No articles found for '{category}'"
+
+        loader = WebBaseLoader(links)
+        docs = loader.load()
+
+        summarization_chain = load_summarize_chain(llm=llm, verbose=False,chain_type='map_reduce')
+        summary = summarization_chain.invoke(docs)
+        return summary
+    except Exception as e:
+        logging.exception(f"Error in Search_Articles for category '{category}': {str(e)}")
+        return f"An error occurred while searching articles for '{category}'. Please try again later."
+
+tools = [
+    StructuredTool.from_function(
```

Comment:
Good use of `StructuredTool` as it is up-to-date with the LangChain configuration.